### ISSUE TITLE
rust should able to infer `Array<T>[CONST_START..CONST_END]` to type `[T; CONST_END - CONST_START]` at compile time

### ISSUE THREAD

**Issue by loynoir** [rust-lang/rust#117599]
### title

rust should able to infer `Array<T>[CONST_START..CONST_END]` to type `[T; CONST_END - CONST_START]` at compile time

### code

```rust
const A: usize = 2;
const B: usize = 6;

fn main() {
    let buf: &[u8] = &[0, 1, 2, 3, 4, 5, 6, 7, 8, 9];

    // OK as expected
    {
        let _: &[u8] = &buf[A..B];
    }

    {
        // // error[E0308]: mismatched types
        // // expected `&[u8; 4]`, found `&[u8]`
        // let _: &[u8; 4] = &buf[A..B];
    }

    // https://doc.rust-lang.org/src/core/array/mod.rs.html
    {
        let _ = <[u8; 4]>::try_from(&buf[A..B]).unwrap();
    }

    // https://docs.rs/crate/arrayref/latest/source/src/lib.rs
    {
        let _ = unsafe { &*((&buf[A..B]).as_ptr() as *const [u8; B - A]) };
    }
}
```

**Comment by wwylele** [comment]
No need to involve `const`. Just `let _: &[u8; 4] = &buf[2..6];` is enough to trigger the error.

To me this is expected. `slice[range]` always returns a &slice. Rust does not have a type system that supports returning different type (either a &slice or an array) depending on the constness of the expression.

**Comment by JulianKnodt** [comment]
I don't believe this is possible to know at compile, because if the input slice does not have enough elements and/or the indexing is out of bounds, then it's not possible to construct an array of the correct size.

In theory if an array is const, and the indexing is const, then it would be theoretically possible, but @wwylele's comment then is applicable.

**Comment by LunarLambda** [comment]
if the slice operation would be out of bounds it will panic and it will never be possible to observe the uninitialized value

I don't see why the compiler couldn't handle this limited sort of casting (similar to integer literals being able to take any fitting type)

It does approach the dependent types territory, a little.
